ee/clickhouse/queries/column_optimizer.py

```diff
@@ -0,0 +1,106 @@
+from typing import List, Set, Tuple, cast
+
+from ee.clickhouse.materialized_columns.columns import ColumnName, get_materialized_columns
+from ee.clickhouse.models.action import get_action_tables_and_properties, uses_elements_chain
+from ee.clickhouse.models.property import extract_tables_and_properties
+from posthog.constants import TREND_FILTER_TYPE_ACTIONS
+from posthog.models.entity import Entity
+from posthog.models.filters import Filter
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.models.property import Property, PropertyName, PropertyType
+from posthog.models.team import Team
+
+
+class ColumnOptimizer:
+    """
+    This class is responsible for figuring out what columns can and should be materialized based on the query filter.
+
+    This speeds up queries since clickhouse ends up selecting less data.
+    """
+
+    def __init__(self, filter: Filter, team_id: int):
+        self.filter = filter
+        self.team_id = team_id
+
+    @cached_property
+    def materialized_event_columns_to_query(self) -> List[ColumnName]:
+        "Returns a list of event table columns containing materialized properties that this query needs"
+
+        materialized_columns = get_materialized_columns("events")
+        return [
+            materialized_columns[property_name]
+            for property_name, type in self._used_properties_with_type("event")
+            if property_name in materialized_columns
+        ]
+
+    @cached_property
+    def should_query_event_properties_column(self) -> bool:
+        return len(self.materialized_event_columns_to_query) != len(self._used_properties_with_type("event"))
+
+    @cached_property
+    def should_query_elements_chain_column(self) -> bool:
+        "Returns whether this query uses elements_chain"
+        has_element_type_property = lambda properties: any(prop.type == "element" for prop in properties)
+
+        if has_element_type_property(self.filter.properties):
+            return True
+
+        if self.filter.filter_test_accounts:
+            test_account_filters = Team.objects.only("test_account_filters").get(id=self.team_id).test_account_filters
+            properties = [Property(**prop) for prop in test_account_filters]
+            if has_element_type_property(properties):
+                return True
+
+        # Both entities and funnel exclusions can contain nested elements_chain inclusions
```

Comment:
@Twixes added some comments around less intuitive bits. Was this what you had in mind?

Reply:
Yup, this is pretty great